streams/src/main/java/org/apache/kafka/streams/processor/internals/RepartitionTopics.java

```diff
@@ -97,49 +101,74 @@ public boolean setup() {
                 }
             }
         }
+    }
 
-        return missingUserInputTopicsPerTopology.isEmpty();
+    public Set<String> topologiesWithMissingInputTopics() {
+        return missingInputTopicsBySubtopology.keySet()
+            .stream()
+            .map(s -> getTopologyNameOrElseUnnamed(s.namedTopology))
+            .collect(Collectors.toSet());
     }
 
-    public Map<String, Set<String>> missingUserInputTopicsPerTopology() {
-        return Collections.unmodifiableMap(missingUserInputTopicsPerTopology);
+    public Queue<StreamsException> missingSourceTopicExceptions() {
+        return missingInputTopicsBySubtopology.entrySet().stream().map(entry -> {
+            final Set<String> missingSourceTopics = entry.getValue();
+            final int subtopologyId = entry.getKey().nodeGroupId;
+            final String topologyName = entry.getKey().namedTopology;
+
+            return new StreamsException(
+                new MissingSourceTopicException(String.format(
+                    "Missing source topics %s for subtopology %s of topology %s",
+                    missingSourceTopics, subtopologyId, topologyName)),
+                new TaskId(subtopologyId, 0, topologyName));
```

Comment:
My point was that I could not see where you use the task ID in this specific case when you the exception is handled and a `StreamsException` can also be created without a task ID. So, I did not see the reason to introduce a dummy task ID if it is not strictly needed by the exception. I am also fine with your current solution.

Reply:
We want to add a TaskId because we want to provide information about which topology is missing the source topics, and to a lesser extent which subtopology. If we create the StreamsException without a TaskId then it's considerably more difficult to figure out which topology (or which query) is experiencing the issue -- technically you could parse the error message in the MissingSourceTopicException but that's obviously very brittle and just not a good idea. Hence, we provide the `topologyName` through a "dummy" TaskId

Reply:
Now I see that that information might be useful in the user-specified exception handler. Thank you for the clarification! Maybe we should use `-1` instead of `0` for the partition number, since the partition number doe not contain any information. Just an idea, not necessarily needed for merging.